Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -97,24 +97,36 @@ extension Array: BSONValue {
             preconditionFailure("Failed to cast values for document \(arrDoc) to array")
         }
 
-       self = arr
+       return arr
     }
 
     public func encode(to storage: DocumentStorage, forKey key: String) throws {
-        // An array is just a document with keys '0', '1', etc. corresponding to indexes
-        if self as? [BSONValue?] == nil {
-            throw MongoError.invalidArgument(message: "Cannot encode a non-BSONValue array element")
-        }
         var arr = Document()
         for (i, v) in self.enumerated() {
-            try arr.setValue(for: String(i), to: v as? BSONValue)
+            guard let val = v as? BSONValue else {
+                throw MongoError.invalidArgument(message: "Cannot encode a non-BSONValue array element")
```

Comment:
Is this still necessary because we don't have conditional conformance before Swift 4.2 and can't ensure that the Array contains only BSONValue instances?

Reply:
Yep! This goes away once we can guarantee conditional conformance to `BSONValue` for arrays of only `[BSONValue]`!!!

Reply:
can we spice up this error message a bit with info about the offending array element? i.e. index, type, string rep

Reply:
We absolute can! This is something we couldn't do before when we threw this on attempt to converting the array itself to `[BSONValue?]`!

Done.